### ISSUE TITLE
Support TRX output file format

### ISSUE THREAD

**Issue by tapika** [xunit/xunit#2188]
To my best understanding xunit has some built-in tools, which are capable of running xunit tests directly - like from command line without using VSTest `vstest.console.exe` ?

Would it be possible to recode test runner to replace `vstest.console.exe` completely ?

`vstest.console.exe` is rather difficult as it does not runs tests directly, but instead it runs testhost, which in a turn runs tests.
This makes rather difficult to debug your own tests when `vstest.console.exe` is executed.

Also I would like to expand support to be able to stop on first exception, instead of continuing running the test.

Brief analysis of `vstest.console.exe` functionality versus azure devops tells me that it's possible also to create a custom command line executable and fully override `vstest.console.exe` behavior. This can be done by responding to `/ListFullyQualifiedTests` and similar kind of command line arguments from here:
https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/VsTestV2/vstest.ts#L330

To be able to list tests in assembly.

And to respond to plain test execution, for example:

`vstest.console mydll.dll /logger:trx`.

Used by azure devops  - command line:
https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/VsTestV2/vstest.ts#L155

Used by azure devops  - usage of trx to publish the results:
https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/VsTestV2/vstest.ts#L638

Test console should be able to generate trx xml file, for which it's possible for example to reuse functionality from trx2unit tool (MIT license), https://github.com/gfoidl/trx2junit/blob/master/source/trx2junit/Internal/junit2trx/TrxTestResultXmlBuilder.cs

Basically by replacing Microsoft tool it would be possible to simplify tests which runs better in production and also expand tool to support new requirements later on.

**Comment by bradwilson** [comment]
> Would it be possible to recode test runner to replace `vstest.console.exe` completely ?

It depends on what you mean by "replace". We can't replace their executable on disk, but in theory someone could write a new one with the same command line switches that only ran xUnit.net tests, and put that in your `PATH` before the real one (or you could invoke it explicitly by path).

That said, a lot of what you're asking for we already support via our own executable. It's unlikely that me, as a single developer, would have the time or desire to add yet another console runner with yet another set of command line switches. If this is really needed, then a third party is going to have to the work (to both create it and support it).

One thing that could make this easier would be to add support in our test runners for TRX output. I think this is something we would consider doing, if you open a new issue for it.

**Comment by tapika** [comment]
What I've briefly analyzed - in azure devops there exists `PublishTestResults` with `testResultsFormat: 'XUnit'` - so in theory it's possible to run xunit test separately and then publish test results - of course it's bit more complex than just to run test and publish test results, but it's doable indeed.

I agree that support for TRX output could be added indeed - that could be first step.
Command line arguments harmonization between xunit runner versus vstest.console.exe should not be difficult after that.
That could be second step. vstest.console vs xunit command line arguments harmonization could happen in parallel as new functionality gets implemented.

Does it makes any sense if we just convert this ticket to 'Add support for TRX output reporter', instead of creating new ticket and closing this one. This however is minor issue, I have some other ideas, which needs to be addressed, but I'll create them as a separate tickets.

**Comment by bradwilson** [comment]
Yep, that'd be fine! 😄

**Comment by bradwilson** [comment]
Note: TRX files can only describe a single test assembly, whereas we support composite result files; therefore, the TRX transformation will have to auto-generate filenames to support multiple assemblies. When results are being output for a single file, the original filename will be used as-is; for everything else, filenames will have `-1`, `-2`, etc. appended to them.

**Comment by bradwilson** [comment]
Fixed in 653218c4f84714e7031c8118d9d0bb1f5529958a
